Encryption.md

```diff
@@ -75,55 +75,126 @@ written to the output stream.
 bottleneck in certain workloads. AES-CTR is a regular (not authenticated) cipher.
 It is faster than AES-GCM, since it doesn’t perform integrity verification and doesn’t 
-calculate the authentication tag. For applications running without AES acceleration and 
+calculate the authentication tag. Actually, AES-GCM is a combination of AES-CTR and an 
+authentication module called GMAC. For applications running without AES acceleration and 
 willing to compromise on content verification, AES-CTR can provide a boost in Parquet 
 write/read throughput. The second Parquet algorithm encrypts the data content (pages) 
 with AES-CTR, and the metadata (Thrift structures) with AES-GCM. This allows to encrypt/decrypt 
 the bulk of the data faster, while still verifying the metadata integrity and making sure 
-the file had not been replaced with a wrong version. However, tampering with the page data 
+the file has not been replaced with a wrong version. However, tampering with the page data 
 might go unnoticed. 
 
-The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
-be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
-length of `aad_metadata` is 512 bytes.
-
-Parquet-mr/-cpp implementations use the RBG-based IV construction as defined in the NIST 
-SP 800-38D document for the GCM ciphers (section 8.2.2).
+The initialization vectors (IVs) in AES GCM and CTR ciphers must be unique for each encryption key, 
+since a single re-use of an IV breaks the encryption. An IV in these two modes is often called a 
+nonce (“number used once”). Parquet encryption uses the RBG-based (random bit generator) IV 
+construction as defined in the NIST SP 800-38D document for the GCM ciphers (section 8.2.2). 
 
 
 ### AES_GCM_V1
-All modules are encrypted with the AES-GCM cipher. The authentication tags (16 bytes) are 
-written after each ciphertext. The IVs (12 bytes) are written before each ciphertext.
+All modules are encrypted by an AES-GCM cipher, without padding. Unique IVs are generated and 
+written before each ciphertext. The IV length is 12 
+bytes (96 bits), which is generally considered to be the optimal size of GCM IVs.  The NIST 
+specification requires a “random field” to fill all 12 bytes and an arbitrary “free field” to be empty for 
+this IV length (see section 8.2.2). 
+
+The authentication tags (16 bytes) are written after each ciphertext.
 
 ### AES_GCM_CTR_V1
-Thrift modules are encrypted with the AES-GCM cipher, as described above. 
-The pages are encrypted with AES-CTR, where the IVs (16 bytes) are written before each 
-ciphertext.
+Thrift modules are encrypted with AES-GCM, as described above. The pages are encrypted by an AES-CTR
+cipher without padding. The nonces are generated and written before each ciphertext, with the nonce length of 16 bytes 
+(128 bits). At least 12 bytes out of 16 are constructed with an RBG.
```

Comment:
Why are nonces and IVs different lengths? Why can't a 12-byte IV that is entirely generated with an RBG be used for CTR mode?

Reply:
The links don't address my question. AES-GCM requires a 12-byte IV, to which a 4-byte counter is added to get the nonce for AES-CTR. Why doesn't the counter mode do the same and take an IV to which it adds the counter? Is this a quirk of the API for the AES-CTR library you're using, or is there a real difference between how the two are handled?

Reply:
It works like this:

1. **AES CTR**
requires a user to provide a 16 byte IV for each encryption operation [API call to convert n-byte plaintext into n-byte ciphertext - takes the following input parameters: key (16 bytes), IV (16 bytes), plaintext offset, and length n].

Per the quotes above, the IV can be constructed by the user in one of the two following ways:
a. First 12 bytes are random bytes, the other 4 are set to a value of 1.
b. All 16 bytes are random.

Inside, CTR implementation will use the last 4 bytes as a counter field when running over n bytes split into 16 byte chunks.

2. **AES GCM**
requires a user to provide a 12 byte IV (see NIST spec) for each encryption operation [API call to convert n-byte plaintext into n-byte ciphertext and 16-byte tag - takes the following input parameters: key (16 bytes), IV (12 bytes), plaintext offset, and length n].

Per the NIST spec section 8.2.2, the IV is constructed by the user as a 12-byte random array.

Inside, GCM runs a CTR cipher that gets a 16-byte IV, constructed from the 12-byte array provided by user, and extra 4 bytes provided by the GCM implementation itself. The CTR cipher then runs as described above ("will use the last 4 bytes as a counter field when running over n bytes split into 16 byte chunks").

Reply:
Okay, then it sounds like the problem is just with the API that is used. One accepts a 12-byte IV and constructs a 16-byte IV+counter, the other accepts a 16-byte nonce where 12 bytes are an IV and it overwrites some bytes with the counter.

For Parquet, we don't need two formats. The encryption library should use the same 12-byte IV to construct a nonce that is acceptable for passing into AES-CTR if that API is different.